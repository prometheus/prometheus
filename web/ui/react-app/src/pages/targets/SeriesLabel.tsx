import * as React from 'react';
import { Badge } from 'reactstrap';
import styles from './SeriesLabel.module.css';

export interface SeriesLabelProps {
  labelKey: string;
  labelValue: string;
}

const SeriesLabel: React.FC<SeriesLabelProps> = ({ labelKey, labelValue }) => {
  return (
    <Badge color="primary" className={styles['series-label']}>
      {`${labelKey}="${labelValue}"`}
    </Badge>
  );
};

export default SeriesLabel;
