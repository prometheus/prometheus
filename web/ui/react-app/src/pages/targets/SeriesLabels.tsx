import * as React from 'react';
import { Tooltip } from 'reactstrap';
import SeriesLabel from './SeriesLabel';

interface Labels {
  [key: string]: string;
}

export interface SeriesLabelsProps {
  discoveredLabels: Labels;
  labels: Labels;
}

const formatLabels = (labels: Labels): string[] => Object.keys(labels).map(key => `${key}="${labels[key]}"`);

const SeriesLabels: React.FC<SeriesLabelsProps> = ({ discoveredLabels, labels }: SeriesLabelsProps) => {
  const [tooltipOpen, setTooltipOpen] = React.useState(false);

  const toggle = (): void => setTooltipOpen(!tooltipOpen);

  return (
    <>
      <div id="series-labels" className="series-labels-container">
        {Object.keys(labels).map(key => {
          return <SeriesLabel key={key} labelKey={key} labelValue={labels[key]} />;
        })}
      </div>
      <Tooltip isOpen={tooltipOpen} target="series-labels" toggle={toggle}>
        <b>Before relabeling:</b>
        {formatLabels(discoveredLabels).map((s: string, idx: number) => (
          <React.Fragment key={idx}>
            <br />
            <span>{s}</span>
          </React.Fragment>
        ))}
      </Tooltip>
    </>
  );
};

export default SeriesLabels;
