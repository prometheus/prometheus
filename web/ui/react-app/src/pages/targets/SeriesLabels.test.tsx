/* eslint @typescript-eslint/camelcase: 0 */

import * as React from 'react';
import { shallow } from 'enzyme';
import SeriesLabels from './SeriesLabels';
import { Tooltip } from 'reactstrap';
import SeriesLabel from './SeriesLabel';
import toJson from 'enzyme-to-json';

describe('SeriesLabels', () => {
  const defaultProps = {
    discoveredLabels: {
      __address__: 'localhost:9100',
      __metrics_path__: '/metrics',
      __scheme__: 'http',
      job: 'node_exporter',
    },
    labels: {
      instance: 'localhost:9100',
      job: 'node_exporter',
      foo: 'bar',
    },
  };
  const seriesLabels = shallow(<SeriesLabels {...defaultProps} />);

  it('renders a div of series labels', () => {
    const div = seriesLabels.find('div').filterWhere(elem => elem.hasClass('series-labels-container'));
    expect(div).toHaveLength(1);
    expect(div.prop('id')).toEqual('series-labels');
  });

  it('wraps each label in a SeriesLabel', () => {
    const l: { [key: string]: string } = defaultProps.labels;
    Object.keys(l).forEach((key: string): void => {
      const label = seriesLabels.find(SeriesLabel).filterWhere(label => label.prop('labelKey') === key);
      expect(label.prop('labelValue')).toEqual(l[key]);
    });
    expect(seriesLabels.find(SeriesLabel)).toHaveLength(3);
  });

  it('renders a tooltip for discovered labels', () => {
    const tooltip = seriesLabels.find(Tooltip);
    expect(tooltip).toHaveLength(1);
    expect(tooltip.prop('isOpen')).toBe(false);
    expect(tooltip.prop('target')).toBe('series-labels');
    expect(tooltip.prop('isOpen')).toBe(false);
  });

  it('renders discovered labels', () => {
    expect(toJson(seriesLabels)).toMatchSnapshot();
  });
});
