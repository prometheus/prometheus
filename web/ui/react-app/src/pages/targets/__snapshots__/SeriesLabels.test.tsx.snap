// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesLabels renders discovered labels 1`] = `
<Fragment>
  <div
    className="series-labels-container"
    id="series-labels"
  >
    <SeriesLabel
      key="instance"
      labelKey="instance"
      labelValue="localhost:9100"
    />
    <SeriesLabel
      key="job"
      labelKey="job"
      labelValue="node_exporter"
    />
    <SeriesLabel
      key="foo"
      labelKey="foo"
      labelValue="bar"
    />
  </div>
  <Tooltip
    autohide={true}
    isOpen={false}
    placement="top"
    placementPrefix="bs-tooltip"
    target="series-labels"
    toggle={[Function]}
    trigger="click hover focus"
  >
    <b>
      Before relabeling:
    </b>
    <br />
    <span>
      __address__="localhost:9100"
    </span>
    <br />
    <span>
      __metrics_path__="/metrics"
    </span>
    <br />
    <span>
      __scheme__="http"
    </span>
    <br />
    <span>
      job="node_exporter"
    </span>
  </Tooltip>
</Fragment>
`;
